spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala

```diff
@@ -0,0 +1,376 @@
+        partitionFileCounter.put(partition.##, addSub)
+    }
+  }
+
+  /** Get the maximum number of files among all partitions inside table `tableId`. */
+  def maxNumFilesInTable(tableId: String): Long = {
+    synchronized {
+      tablePartitionStatsCache.get(tableId) match {
+        case Some(partitionFileCounter) =>
+          if (partitionFileCounter.isEmpty) {
+            0
+          } else {
+            partitionFileCounter.map(_._2.numFiles).max
+          }
+        case None => 0
+      }
+    }
+  }
+
+  /**
+   * @return Filter partitions from targetPartitions that have not been auto-compacted or
+   *         that have enough small files.
+   */
+  def filterPartitionsWithSmallFiles(tableId: String, targetPartitions: Set[PartitionKey],
+      minNumFiles: Long): Set[PartitionKey] = synchronized {
+    tablePartitionStatsCache.get(tableId).map { tablePartitionStates =>
+      targetPartitions.filter { partitionKey =>
+        tablePartitionStates.get(partitionKey.##).exists { partitionState =>
+          partitionState.hasSufficientSmallFilesOrHasNotBeenCompacted(minNumFiles)
+        }
+      }
+    }.getOrElse(Set.empty)
+  }
+
+  def markPartitionsAsCompacted(tableId: String, compactedPartitions: Set[PartitionKey])
+  : Unit = synchronized {
+    tablePartitionStatsCache.get(tableId).foreach { tablePartitionStats =>
+      compactedPartitions
+        .foreach(partitionKey => tablePartitionStats.get(partitionKey.##)
+          .foreach(_.wasAutoCompacted = true))
+    }
+  }
+
+  /**
+   * Collect the number of files, which are less than minFileSize, added to or removed from each
+   * partition from `actions`.
+   */
+  def collectPartitionStats(
+      collector: AutoCompactPartitionStatsCollector,
+      tableId: String,
+      actions: Iterator[Action]): Unit = {
+    val acts = actions.toVector
+    acts.foreach {
+      case addFile: AddFile => collector.collectPartitionStatsForAdd(addFile)
+      case removeFile: RemoveFile => collector.collectPartitionStatsForRemove(removeFile)
+      case _ => // do nothing
+    }
+    collector.finalizeStats(tableId)
+  }
+
+  /** This is test only code to reset the state of table partition statistics. */
+  private[delta] def resetTestOnly(newHashSpace: Int, newMaxNumPartitions: Int): Unit = {
+    synchronized {
+      tablePartitionStatsCache.clear()
+      maxNumTablePartitions = newHashSpace
+      maxNumPartitions = newMaxNumPartitions
+      numUsedPartitions = 0
+      AutoCompactPartitionReserve.resetTestOnly()
+    }
+  }
+
+  /**
+   * This is test only code to reset all partition statistic information and keep current
+   * configuration.
+   */
+  private[delta] def resetTestOnly(): Unit = resetTestOnly(maxNumTablePartitions, maxNumPartitions)
+}
+
+object AutoCompactPartitionStats {
+  private var _instance: AutoCompactPartitionStats = null
+
+  /** The thread safe constructor of singleton. */
+  def instance(spark: SparkSession): AutoCompactPartitionStats = {
```

Comment:
Why synchronized? Why is there a need for a singleton?

Reply:
Multiple threads could be committing simultaneously, but we want to keep all the stats together. This gives us a safe way to have a single source of stats across all transactions